org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/access/binary/BinaryClass.java

```diff
@@ -228,6 +230,35 @@ public Throwable fillInStackTrace() {
 			}
 		};
 	}
+	
+	private static final boolean IS_CASE_SENSITIVE = isCaseSensitive();
+	
+	private static boolean isCaseSensitive() {
+		String result = System.getProperty("org.eclipse.xtext.common.types.access.binary.BinaryClass.ASSUME_CASE_SENSITIVE", Boolean.toString(inferIsCaseSensitive()));
+		return Boolean.parseBoolean(result);
+	}
+	
+	private static boolean inferIsCaseSensitive() {
+		try {
+			File tempFile = File.createTempFile("prefix", "");
+			File upperCase = new File(tempFile.getParentFile(), tempFile.getName().toUpperCase());
+			return !(upperCase.exists() && upperCase.getCanonicalFile().getName().equals(tempFile.getName()));
+		} catch(IOException ex) {
+			return false;
+		}
+	}
+	
+	private static boolean fileSystemAwareExists(URL url) {
+		if (!IS_CASE_SENSITIVE && "file".equals(url.getProtocol())) {
```

Comment:
do we have estimates how the penalty for this is

Reply:
Unfortunately not. I’m afraid it’s expensive though. That’s why I introduced a means to skip the check.